File: neetcode-kotlin/app/src/main/kotlin/com/niyaj/neetcode/MergeTwoSortedLinkedLists.kt
package com.niyaj.neetcode

/**
 * Merge Two Sorted Linked Lists
 * You are given the heads of two sorted linked lists list1 and list2.
 *
 * Merge the two lists into one sorted linked list and return the head of the new sorted linked list.
 *
 * The new list should be made up of nodes from list1 and list2.
 *
 * Example 1:
 *
 *
 *
 * Input: list1 = [1,2,4], list2 = [1,3,5]
 *
 * Output: [1,1,2,3,4,5]
 * Example 2:
 *
 * Input: list1 = [], list2 = [1,2]
 *
 * Output: [1,2]
 * Example 3:
 *
 * Input: list1 = [], list2 = []
 *
 * Output: []
 * Constraints:
 *
 * 0 <= The length of the each list <= 100.
 * -100 <= Node.val <= 100
 */
object MergeTwoSortedLinkedLists {
    fun mergeTwoLists(list1: ListNode?, list2: ListNode?): ListNode? {
        if (list1 == null) return list2
        if (list2 == null) return list1

        if (list1.value <= list2.value) {
            list1.next = mergeTwoLists(list1.next, list2)
            return list1;
        } else {
            list2.next = mergeTwoLists(list1, list2.next)
            return list2
        }
    }
}

fun main() {
    val list1 = ListNode(1, ListNode(2, ListNode(4)))
    val list2 = ListNode(1, ListNode(3, ListNode(5)))

    MergeTwoSortedLinkedLists.mergeTwoLists(list1, list2).printList()
}

File: neetcode-cpp/MergeTwoSortedLinkedLists.cpp
#include <iostream>

#include "ListNode.h"
using namespace std;

class MergeTwoSortedLinkedLists {
public:
    static auto mergeTwoLists(ListNode *list1, ListNode *list2) -> ListNode *;
};
ListNode *MergeTwoSortedLinkedLists::mergeTwoLists(ListNode *list1, ListNode *list2) {
    if (!list1)
        return list2;
    if (!list2)
        return list1;

    if (list1->val <= list2->val) {
        list1->next = mergeTwoLists(list1->next, list2);
        return list1;
    } else {
        list2->next = mergeTwoLists(list1, list2->next);
        return list2;
    }
}

int main() {
    auto *list1 = new ListNode(1);
    list1->next = new ListNode(2);
    list1->next->next = new ListNode(4);

    auto *list2 = new ListNode(1);
    list2->next = new ListNode(3);
    list2->next->next = new ListNode(4);

    ListNode *result = MergeTwoSortedLinkedLists::mergeTwoLists(list1, list2);

    ListNode::printList(result);

    return 0;
}
